### ISSUE TITLE
Feature Request: Variable Capture from Block Outputs

### ISSUE THREAD

**Issue by gouravroy2110** [srinjoy-2005/ArchIDE#25]
## Summary
In dynamic or complex architectures, intermediate tensor shapes depend on runtime outputs of preceding blocks. Currently, variables in `VariablesPanel` are strictly static constants defined by the user.

## Proposed Solution
- **Variable Binding**: Allow variables defined in `VariablesPanel` to bind to and capture the output tensor or shape dimensions of a specific node on the canvas.
- **Canvas Interaction**: Provide a "Capture Output to Variable" option on block output ports or the Properties panel.
- **Compiler Integration**: The shape inference pass and code generator can propagate these captured variable symbols to downstream blocks (e.g. via `@var:<var_name>`) for dynamic reshape or dimension calculations.

**Comment by gouravroy2110** [comment]
## Developer Implementation Guide: Variable Capture from Block Outputs

### 1. Overview
In modern dynamic architectures, downstream tensor dimensions and configurations often depend on runtime outputs of preceding layers (e.g. sequence length after tokenization, extracted feature channels, or flattened dimensions). This feature enables binding variables in the Variables panel directly to block outputs or tensor shape dimensions on the canvas.

### 2. Data Model Specification

Extend `VariableDefinition` in `src/lib/types.ts` and `backend/models.py`:

```typescript
export interface VariableDefinition {
  id: string;
  name: string;
  type: 'int' | 'float' | 'string' | 'tuple';
  default_value: string | number;
  // New binding specification:
  captured_from?: {
    node_id: string;        // UUID or alias of the source node on canvas
    port_id: string;        // Output port ID (e.g. 'out', 'shape', 'dim_1')
    dim_index?: number;     // Optional dimension index (e.g. 1 for C in (B, C, H, W))
  };
}
```

### 3. UI Workflow & Interactions

```
+-------------------------------------------------------------+
| Variables Panel                                        [+]  |
+-------------------------------------------------------------+
| Name        | Type  | Value / Source                        |
+-------------+-------+---------------------------------------+
| batch_size  | int   | 1                                     |
| hidden_dim  | int   | ⚡ Captured: linear_1.out [dim: -1]    |
+-------------------------------------------------------------+
```

1. **Binding via Variables Panel**:
   - In `VariablesPanel.tsx`, each variable row has a "Capture Source" dropdown.
   - Selecting "Capture from Node" displays a picker of nodes and their output ports, with an optional dimension index selector (e.g. `dim 0`, `dim 1`, `dim 2`, `dim 3`, or `full shape`).
2. **Binding via Canvas Port**:
   - Right-clicking an output port handle on canvas -> "Capture to Variable...".
   - Opens quick dialog to create a new variable or bind to an existing one.
3. **Canvas Visual Cue**:
   - Ports with active variable bindings display a small lightning icon or badge: `⚡ @var:hidden_dim`.

### 4. Compiler & Code Generation

During topological code generation in `backend/compiler.py`:

1. When generating code for a node that has captured variables:
   ```python
   # Source node forward output:
   out_1 = self.linear_1(x)
   # Captured variable code emission:
   hidden_dim = out_1.shape[-1]
   ```
2. Downstream nodes referencing `@var:hidden_dim`:
   - In forward pass, downstream operations reference the local variable `hidden_dim` instead of a static constructor parameter.
   - If the downstream block is an `nn.Module` requiring this dimension in `__init__`, the compiler reports a diagnostic warning explaining that runtime-captured variables can only be used in functional operations or lazy modules (`nn.LazyLinear`, `nn.LazyConv2d`).

### 5. Implementation Tasks
- [ ] Update `VariableDefinition` in `src/lib/types.ts` and `backend/models.py`.
- [ ] Add variable capture binding UI in `src/components/VariablesPanel.tsx`.
- [ ] Add context menu or port click action in `src/components/CustomNode.tsx` for capturing ports.
- [ ] Update `backend/compiler.py` forward pass emitter to synthesize `.shape[...]` variable assignments.
- [ ] Add backend unit tests in `backend/tests/test_variable_capture.py`.
